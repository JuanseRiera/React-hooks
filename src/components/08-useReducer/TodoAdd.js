import React from "react";
import PropTypes from "prop-types";
import { useForm } from "../../Hooks/useForm";

export const TodoAdd = ({ guardarTarea }) => {
  const { formState, changeInput, reset } = useForm({
    todoInput: "",
  });

  const submit = (e) => {
    e.preventDefault();
    guardarTarea({
      id: Math.round(Math.random() * 100000),
      todo: formState.todoInput,
      done: false,
    });
    reset();
  };
  return (
    <>
      <form onSubmit={submit} className="text-end">
        <input
          type="text"
          className="form-control"
          placeholder="Ingrese su tarea por hacer..."
          name="todoInput"
          autoComplete="off"
          onChange={changeInput}
          value={formState.todoInput}
        />
      </form>
    </>
  );
};

TodoAdd.propTypes = {
  guardarTarea: PropTypes.func.isRequired,
};
