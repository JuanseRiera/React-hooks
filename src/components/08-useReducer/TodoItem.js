import React from "react";
import PropTypes from "prop-types";

export const TodoItem = ({ item, toggle, eliminar }) => {
  const { id, todo, done } = JSON.parse(item);
  return (
    <>
      <div className="card mb-3">
        <div className="card-body row">
          <div className="col-12 col-sm-8 my-auto">
            <h6
              className={done ? "text-decoration-line-through" : null}
              onClick={() => {
                toggle(id);
              }}
            >
              {todo}
            </h6>
          </div>
          <div className="col-12 col-sm-4 text-end">
            <button
              className="btn btn-danger ms-2"
              onClick={() => {
                eliminar(id);
              }}
            >
              Borrar
            </button>
          </div>
        </div>
      </div>
    </>
  );
};

TodoItem.propTypes = {
  item: PropTypes.string.isRequired,
  toggle: PropTypes.func.isRequired,
  eliminar: PropTypes.func.isRequired,
};
