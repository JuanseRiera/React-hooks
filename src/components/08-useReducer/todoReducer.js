export const todoReducer = (state, action) => {
  switch (action.type) {
    case "agregar":
      return [...state, action.payload];

    case "toggle":
      return state.map((todo) => {
        if (todo.id === action.payload) {
          return {
            ...todo,
            done: !todo.done,
          };
        } else {
          return todo;
        }
      });

    case "eliminar":
      return state.filter((x) => x.id !== action.payload);

    default:
      return state;
  }
};
