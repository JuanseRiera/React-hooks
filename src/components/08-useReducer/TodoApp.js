import React, { useEffect, useReducer } from "react";
import { TodoItem } from "./TodoItem";
import { todoReducer } from "./todoReducer";
import "./reducer.css";
import { TodoAdd } from "./TodoAdd";

export const TodoApp = () => {
  const init = () => {
    return JSON.parse(localStorage.getItem("tareas")) || [];
  };

  const [todos, dispatch] = useReducer(todoReducer, [], init);

  const guardarTarea = (todo) => {
    let actionAgregar = {
      type: "agregar",
      payload: todo,
    };
    dispatch(actionAgregar);
  };

  const toggle = (id) => {
    let actiontoggle = {
      type: "toggle",
      payload: id,
    };
    dispatch(actiontoggle);
  };

  const eliminar = (id) => {
    let actionEliminar = {
      type: "eliminar",
      payload: id,
    };
    dispatch(actionEliminar);
  };

  useEffect(() => {
    localStorage.setItem("tareas", JSON.stringify(todos));
  }, [todos]);

  return (
    <>
      <div className="row">
        <div className="col-12 col-md-4 my-auto">
          <h1>TodoApp</h1>
        </div>
        <div className="col-12 col-md-8 my-auto">
          <TodoAdd guardarTarea={guardarTarea} />
        </div>
      </div>

      <hr />

      <div className="row">
        <div className="col-12 p-3">
          {todos.map((todo) => {
            return (
              <TodoItem
                key={todo.id}
                item={JSON.stringify(todo)}
                toggle={toggle}
                eliminar={eliminar}
              />
            );
          })}
        </div>
      </div>
    </>
  );
};
