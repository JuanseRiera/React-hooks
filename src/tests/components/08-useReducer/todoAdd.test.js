import { shallow } from "enzyme";
import { TodoAdd } from "../../../components/08-useReducer/TodoAdd";
describe("Pruebas del todoadd", () => {
  const guardarTarea = jest.fn();
  const wrapper = shallow(<TodoAdd guardarTarea={guardarTarea} />);
  test("Deberia de cargar correctamente", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("NO deberia de llamarse a guardar tarea", () => {
    const onSubmit = wrapper.find("form").prop("onSubmit");
    onSubmit({ preventDefault: jest.fn() });
    expect(guardarTarea).toHaveBeenCalledTimes(0);
  });

  test("Deberia de llamarse a guardar tarea", () => {
    wrapper
      .find("input")
      .simulate("change", { target: { name: "todoInput", value: "Changed" } });
    const onSubmit = wrapper.find("form").prop("onSubmit");
    onSubmit({ preventDefault: jest.fn() });
    expect(guardarTarea).toHaveBeenCalledTimes(1);
    expect(guardarTarea).toHaveBeenLastCalledWith({
      id: expect.any(Number),
      todo: "Changed",
      done: false,
    });
    expect(wrapper.find("input").prop("value")).toBe("");
  });
});
