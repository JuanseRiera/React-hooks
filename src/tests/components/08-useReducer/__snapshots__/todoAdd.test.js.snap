// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas del todoadd Deberia de cargar correctamente 1`] = `
<Fragment>
  <form
    className="text-end"
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="form-control"
      name="todoInput"
      onChange={[Function]}
      placeholder="Ingrese su tarea por hacer..."
      type="text"
      value=""
    />
  </form>
</Fragment>
`;
